docs/03-code-internals/19-service-objects.md

```diff
@@ -0,0 +1,622 @@
+Everything that happens for a business action will be done in the service, let it be by using specialized steps, custom methods or delegating things to other objects.
+
+Our service concepts are heavily inspired by [Trailblazer 2.0](https://trailblazer.to/2.0/), [dry-transaction](https://dry-rb.org/gems/dry-transaction) and to a lesser extent [interactor](https://github.com/collectiveidea/interactor).
+
+## Why?
+
+The most common place where to use a service object is in a controller action. A controller action is usually a point of entry for dedicated business logic, but it’s not always clear where to put that logic. Of course, it can stay in the controller action, that’s what a lot of Rails apps do, but then it tends to grow and repeat the same processes over and over. Sometimes part of that logic can be extracted into models, and that’s fine, but it can lead to what’s called fat models where a model starts handling way more things than it should.
+
+That’s a typical case where a service object comes in handy: you describe what the business logic is step by step (fetching models, checking permissions, and so on) and then you can take action depending on the outcome in a very simple way. All the error handling is done for you inside the service, and you can match its outcome in a very deterministic way. Input parameters are validated using a dedicated object called a contract and when a step fails, the service stops. That means that if a step runs, every previous step was successful.
+
+It doesn’t mean every situation should be handled by a service. A service encapsulates business logic for a given action, and is encouraged to rely on specialized objects. So it won’t replace models or libraries, for example. As said before, it really shines when used in a controller action, but it can be used anywhere. A background job is another good example.
+
+Another benefit immediately available with services, is that since all the logic related to a business action is encapsulated in it, you can call it from anywhere (console, SDK, controllers, specs, etc.) and you’ll always get the same outcome.
+
+It’s also quite easy to understand what’s happening in a service at a glance, since all the steps are listed sequentially. There’s also a matching system to handle the possible outcomes in a deterministic way, here again you can understand at a glance what a controller will do, for example.
+
+## Getting started
+
+Here’s a simplified service to update a user’s username which demonstrates most available steps:
+```ruby
+class User::UpdateUsername
+  include Service::Base
+
+  params do
+    attribute :id, :integer
+    attribute :username, :string
+
+    validates :id, presence: true
+    validates :username, presence: true, format: { with: /\A[a-zA-Z0-9]+\z/ }
+  end
+  model :user
+  policy :can_update_username
+  transaction do
+    step :update
+    step :log
+  end
+
+  private
+
+  def fetch_user(params:)
+    User.find_by(id: params[:id])
```

Comment:
params.id ?

Reply:
related question do we prevent `params[:id] = "something"`, it should be immutable IMO

Reply:
> params.id ?

Now that we’re using `params` instead of `contract`, it seems more consistent to keep using an API most people expect when encountering `params` (since most of the time it will come from a controller anyway). The rationale behind it is to reduce confusion at that level.

> related question do we prevent `params[:id] = "something"`? it should be immutable IMO

I can explore this option, I think we have 2/3 services doing this at the moment, but indeed, if you need processing, it should be done in the contract not outside of it.

Reply:
Yes I ceel pretty stringly it should be immutable

Im fine choosing params[:key] over params.key but options should follow the same, and params.key or options.key should NOT be possible then

Reply:
For immutability, I’m currently checking what are the necessary changes for the existing services.

Regarding having only one way to access things, I’m not sure about that. Being able to do `params.key` is nicer I think, even if it’s less consistent.

Reply:
I just made two PRs (the second one is based on the first one):
- https://github.com/discourse/discourse/pull/29441
- https://github.com/discourse/discourse/pull/29443